### ISSUE TITLE
Add broadcast system for the master node

### ISSUE THREAD

**Issue by AdriiiPRodri** [wazuh/wazuh#12912]
## Description

In this issue, the goal is to create a broadcasting mechanism for the cluster. Currently, there is only one task that is started on the master node and sent to all workers. This task is the synchronization of agent-groups.

There is a WorkerHandler for each connection of a worker to the master, each of these WorkerHandler executes a periodic task of obtaining data that is requested to the master.

The problem with this type of task is that for every iteration, the information must be sent only once to each worker. To achieve this, certain processes have to be added that increase the use of resources.

Therefore in this issue, we will develop a generic system that will help us to improve resource consumption as well as to make this mechanism clearer and more concise. Ideally, it would be that the master node could get the information and it would broadcast it to all WorkerHandlers and they would send it. In this way, the WorkerHandler would not have to have a periodic task to check if there is data to send. This mechanism should work generically for the development of agent-groups and for future developments that need this feature.

## TODO

- [x] Add the new broadcast system.
- [x] Modify unit tests according to change.
- [ ] ~Perform a study of the performance.~ (Should be done in a new issue in the agent-groups epic, #10771)
- [ ] ~Run the system/reliability tests to check that everything is still working as expected.~ (Should be done in a new issue in the agent-groups epic, #10771)

**Comment by Selutario** [comment]
## Status update
### Description
The main goal of this issue is to create a system that allows the master to send a message to all connected workers. For this I am developing a structure like the one you can see in the image below. 

This structure consists of adding an Asyncio Queue as an attribute in the `AbstractServerHandler`. Everytime a worker node connects to the master, an instance of the `MasterHandler` class (which inherits from `AbstractServerHandler`) is created. As a consequence, each master-worker connection will contain an asynchronous queue on the master side.
![schema_1](https://user-images.githubusercontent.com/23361101/162743472-c5963435-aded-4a82-a556-8240ba7a599b.png)

So far I have added the following methods to be used in any future development (their names can still change):
- **broadcast:** It generates an identifier so responses can be stored and adds a request to the queue of each connected client.
- **broadcast_sent:** It checks if the request in each worker node was sent. If all requests were sent, `True` is returned.
- **get_broadcast_result:** It returns a dict with the response from each request and delete the dict where the response was stored.

### Workflow
The desired/expected workflow would be:
1. The **broadcast(...)** method is run. As parameters, it receives a reference to a method defined inside AbstractServerHandler or any inheriting class and the args and kwargs needed for said method. This makes the broadcasting system very flexible and reusable.
2. A request containing the function reference and its parameters is added to the Async Queue of each AbstractServerHandler.
3. Inside each AbstractServerHandler, as soon as the item has been added (and the server is free), the method is run and the response obtained. The method here could be anything. For instance, a function that sends a message to the worker.
4. The response of the method is stored in a dict defined in the AbstractServer where it can be queried by the `broadcast_sent` and `get_broadcast_result` functions, if needed.

**Comment by Selutario** [comment]
## Status update

The development of the broadcasting system is ready and in my preliminary tests everything works correctly. Finally, this would be the interface or functions available for the use of broadcasting in the cluster:
- **broadcast:** Run a function in all server handlers. There is one server handler for each connected client. Nothing is returned and it is not possible to check if the function was run in every server handler or the result.
- **broadcast_add:** It works as the method above (`broadcast`) but it does return an identifier (`broadcast_id`). When this function is used, the result of each server handler is stored and it can be queried with its `broadcast_id`.
- **broadcast_pop:** Receive `broadcast_id` as a parameter. It returns `False` if the request has not yet been run in all expected server handlers. Otherwise, a dict with the response for each one is returned.

## To do
- Update unit tests.
- Performe more manual tests.
- Use the new broadcast system in [this agent-groups development](https://github.com/wazuh/wazuh/issues/10771) and make sure that everything still works.

**Comment by Selutario** [comment]
## Status update

### 1. Cancel broadcast_reader task (done)

An asyncio task in charge of reading any item in the queue was created inside the `hello()` function every time that a worker node connected to the master: 
https://github.com/wazuh/wazuh/blob/5ce9cd8e2ab79e147cfbeb33d911668a6be83903/framework/wazuh/core/cluster/server.py#L151

However, when the worker disconnected and the Handler instance was deleted, said task would not be removed. As a consequence, the number of running tasks would increase each time that a new worker-master connection was established. I changed this so we store the task instance in order to delete it when the connection is lost:
```PYTHON
            ...
            self.handler_tasks.append(self.loop.create_task(self.broadcast_reader()))
            return b'ok', f'Client {self.name} added'.encode()
```

```PYTHON
            ...
            for task in self.handler_tasks:
                task.cancel()
            ...
```

### 2. Unittests warning (in progress)

This message is being printed in the test_server.py unittests:
```
sys:1: RuntimeWarning: coroutine 'AsyncMockMixin._execute_mock_call' was never awaited
Object allocated at (most recent call last):
  File "/usr/lib/python3.9/unittest/mock.py", lineno 1097
    return self._execute_mock_call(*args, **kwargs)
```

I am trying to find the reason and the solution, although at the moment I have not succeeded.
